integration-test/background/click-to-load-facebook.js

```diff
@@ -1,23 +1,67 @@
 const harness = require('../helpers/harness')
+const { logPageRequests } = require('../helpers/requests')
 
 const testSite = 'https://privacy-test-pages.glitch.me/privacy-protections/click-to-load/'
+const facebookDomains = new Set(['facebook.com', 'facebook.net'])
+
 let browser
 let bgPage
 let teardown
 
+function summariseFacebookRequests (requests) {
+    const facebookSDKRedirect = { checked: false, alwaysRedirected: true }
+    // Note: Requests to the SDK are not included in counts.
+    let requestCount = 0
+    let allowCount = 0
+    let blockCount = 0
+
+    for (const request of requests) {
+        if (facebookDomains.has(request.url.hostname.split('.').slice(-2).join('.'))) {
+            if (request.url.pathname === '/en_US/sdk.js') {
+                facebookSDKRedirect.alwaysRedirected = (
+                    facebookSDKRedirect.alwaysRedirected &&
+                    request.status === 'redirected' &&
+                    request.redirectUrl &&
+                    request.redirectUrl.protocol === 'chrome-extension:' &&
+                    request.redirectUrl.pathname.endsWith('/fb-sdk.js')
+                )
+                facebookSDKRedirect.checked = true
+                continue
+            }
+
+            requestCount += 1
+
+            if (request.status === 'blocked') {
+                blockCount += 1
+            } else {
+                // Consider anything not blocked as allowed, so that block tests
+                // don't miss redirected requests or failed requests.
+                allowCount += 1
+            }
+        }
+    }
+
+    return { facebookSDKRedirect, requestCount, allowCount, blockCount }
+}
+
 describe('Test Facebook Click To Load', () => {
     beforeAll(async () => {
         ({ browser, bgPage, teardown } = await harness.setup())
 
+        // Wait for the extension to load its configuration files.
         await bgPage.waitForFunction(
-            () => {
-                console.log('waiting for tds...')
-                return window.dbg && window.dbg.tds && window.dbg.tds.ClickToLoadConfig
-            },
-            { polling: 100, timeout: 10000 }
+            () => (
+                window.dbg &&
+                window.dbg.tds &&
+                !window.dbg.tds.isInstalling &&
+                window.dbg.tds.ClickToLoadConfig &&
+                window.dbg.tds.tds.trackers &&
+                window.dbg.tds.tds &&
+                window.dbg.tds.tds.domains &&
+                window.dbg.tds.tds.domains['facebook.com'] === 'Facebook, Inc.'
```

Comment:
oh boy, that's a lot, do we need all those checks? Can we compress it with some `?.` magic?

Reply:
Not waiting for all the configuration to load turned out to be the cause of a lot of the test flakes. To help reduce the flakes we had an arbitrary X second wait, but that also meant that the tests ran slower than necessary. I agree it's kind of ugly checking so many things here, but it makes the tests run reliably and quickly so it seems worth it.

I will see if I can split this logic out into a helper at least though.

Reply:
(I moved this logic out into the harness, that has already saved duplication and will help for future tests too.)